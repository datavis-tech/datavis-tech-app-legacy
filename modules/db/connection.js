import shareDBClient from 'sharedb/lib/client'
import ReconnectingWebSocket from 'reconnecting-websocket'

// Initialize the singleton connection.
let connection
if (process.browser) {
  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
  const socket = new ReconnectingWebSocket(protocol + window.location.host)
  connection = new shareDBClient.Connection(socket)
}
export default connection
