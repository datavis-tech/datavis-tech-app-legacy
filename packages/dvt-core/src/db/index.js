const { serializeDocument } = require('./serializers')
const accessors = require('./accessors')

module.exports = {
 serializeDocument,
 accessors
}