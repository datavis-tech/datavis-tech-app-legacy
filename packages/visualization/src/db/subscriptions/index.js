export {default as DocumentSubscription} from './documentSubscription'
export {default as DocumentSubscriptions} from './documentSubscriptions'
export {default as ProfileQuerySubscription} from './profileQuerySubscription'
