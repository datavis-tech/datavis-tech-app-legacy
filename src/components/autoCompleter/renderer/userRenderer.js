import { Label } from 'semantic-ui-react'

export default ({ title }) => <Label content={title} />
