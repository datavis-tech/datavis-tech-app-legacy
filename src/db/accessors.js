const get = require('lodash/get')
// TODO split this into separate files for each accessor.

/****************************
 * For any ShareDB document *
 ****************************/

// Gets the data inside a doc,
// returning an empty object if passed undefined as doc.

const data = doc => (doc && doc.data) ? doc.data : {}

// Returns true if the given document has data loaded, false otherwise.
const hasData = doc => !!(doc && doc.data)

// Access the id of a document.
const id = doc => (doc && doc.id) || ''

/**********************************
 * For "vis" and "data" Documents *
 **********************************/

// This accessor returns the references for the given document,
// or an empty array if there are no references.
const references = doc => data(doc).references || []

// Returns the ids of the references for the given document.
const referenceIds = doc => references(doc).map(reference => reference.id)

// Access the title of a document.
const title = doc => data(doc).title || ''

// Access the description of a document.
const description = doc => data(doc).description || ''

// Access the type of a document.
// If type is undefined, treat it as 'vis'.
const type = doc => data(doc).type || 'vis'

// Access the content of a document.
const content = doc => data(doc).content || ''

// Access the owner of a document.
const owner = doc => data(doc).owner || ''

// access the collaborators of a document.
const collaborators = doc => data(doc).collaborators || []

const collaboratorIds = doc => collaborators(doc).map(c => c.id)

// access the id of the document this one was forked from, if any.
const forkedFrom = doc => data(doc).forkedFrom

const isPrivate = doc => data(doc).isPrivate || false

const thumbnail = doc => data(doc).thumbnail

// This function returns true if all references have been loaded.
// references is the array of reference objects stored in the ShareDB document
// referenceDocuments is the array of serialized documents for references.
const allReferencesLoaded = (references, referenceDocuments) => {

  // If there are no references, return true.
  if (references.length === 0) {
    return true
  }

  // If there are references, return true if there are any elements,
  // because DocumentSubscriptions only passes documents after all are loaded.
  return referenceDocuments && referenceDocuments.length > 0
}

// Constructs the "files" object expected by MagicSandbox.js,
// using the file names from `references`, and content from `referenceDocs`.
const files = (references, referenceDocuments) => references
  .reduce((files, {fileName}, i) => {
    files[fileName] = {
      content: referenceDocuments[i].content
    }
    return files
  }, {})

/**********************************
 * For "profile" Documents *
 **********************************/
const profile = doc => (doc && doc.data) ? doc.data : null

/**********************************
 * For ShareDB Ops (operations)
 **********************************/

// Checks if the given op manipulates the content field of the document.
const isContentOp = op => get(op, 'op[0].p[0]') === 'content'

module.exports = {
  hasData,
  id,
  references,
  referenceIds,
  title,
  description,
  type,
  content,
  owner,
  collaborators,
  collaboratorIds,
  forkedFrom,
  isPrivate,
  allReferencesLoaded,
  files,
  profile,
  thumbnail,
  isContentOp
}
