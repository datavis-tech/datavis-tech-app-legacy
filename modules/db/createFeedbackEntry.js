import connection from './connection'
import { DB_FEEDBACK_COLLECTION } from '../constants'
import generateId from './generateId'


// Creates an entry in the feedback collection.
const createFeedbackEntry = ({ feedback, user }) => {
  const id = generateId()
  const doc = connection.get(DB_FEEDBACK_COLLECTION, id)
  doc.create({

    // Human readable feedback, String.
    feedback,

    // The user who left this feedback.
    user
  })
}

export default createFeedbackEntry
