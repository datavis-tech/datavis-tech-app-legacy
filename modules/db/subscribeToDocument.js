import { DB_DOCUMENTS_COLLECTION } from '../constants'
import connection from './connection'

const subscribeToDocument = (id, callback) => {
  const doc = connection.get(DB_DOCUMENTS_COLLECTION, id)
  doc.subscribe((err) => {
    callback(err, doc)
  })
}

export default subscribeToDocument
