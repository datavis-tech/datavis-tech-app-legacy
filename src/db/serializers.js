import * as accessors from './accessors'

export const serializeDocument = (document) => (
  {
    id: accessors.id(document),
    type: accessors.type(document),
    title: accessors.title(document),
    description: accessors.description(document),
    content: accessors.content(document),
    isPrivate: accessors.isPrivate(document),
    references: accessors.references(document),
    referencesIds: accessors.referenceIds(document),
    collaborators: accessors.collaborators(document),
    collaboratorsIds: accessors.collaboratorIds(document),
    forkedFrom: accessors.forkedFrom(document)
  }
)
