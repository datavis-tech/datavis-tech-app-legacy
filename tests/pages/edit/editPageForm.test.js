import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import fakeDoc from '../../utils/fakeDoc'
import nodeSelector from '../../utils/nodeSelector'
import { serializeDocument } from '../../../src/db/serializers'
import CodeMirrorBinding from '../../../src/components/codeMirrorBinding'
import EditPageForm from '../../../src/pages/edit/editPageForm'

describe('edit page form', () => {

  let sut
  let props
  let __shareDbDoc
  let document
  let onTitleChange
  let onPrivacyChange
  let onDescriptionChange

  beforeEach(() => {

    __shareDbDoc = fakeDoc()
    document = serializeDocument(__shareDbDoc)

    document = jest.fn()
    onTitleChange = jest.fn()
    onPrivacyChange = jest.fn()
    onDescriptionChange = jest.fn()

    props = { plan: 'early-adopter', document, __shareDbDoc, onTitleChange, onPrivacyChange, onDescriptionChange }

    sut = shallow(<EditPageForm {...props} />)
  })

  it('should render content as code mirror', () => {
    expect(sut.find(CodeMirrorBinding).props()).toMatchObject({
      doc: __shareDbDoc,
      path: ['content']
    })
  })

  describe('changes of inputs', () => {

    testInput('privacy-public', 'onPrivacyChange', false)
    testInput('privacy-private', 'onPrivacyChange', true)

    function testInput (name, prop, expectedValue) {
      it(`should notify about changes of ${name}`, () => {
        const value = String(Math.random())
        sut.find(nodeSelector(name)).simulate('change', {target: {value}})
        expect(props[prop]).toHaveBeenCalledWith(expectedValue === undefined ? value : expectedValue)
      })
    }
  })

  describe('layouts', () => {

    it('should not contain slots if they were not provided', () => {
      const tree = renderer.create(<EditPageForm {...props} />).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should contain slots if they were provided', () => {

      const propsWithSlots = {
        References: <div>References</div>,
        Preview: <div>Preview</div>,
        ...props
      }

      const tree = renderer.create(<EditPageForm {...propsWithSlots} />).toJSON()
      expect(tree).toMatchSnapshot()
    })

  })

})
