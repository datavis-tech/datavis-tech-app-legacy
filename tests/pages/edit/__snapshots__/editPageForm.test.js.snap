// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit page form layouts should contain slots if they were provided 1`] = `
<form
  action={undefined}
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label>
      Title
    </label>
    <div
      className="ui divided two column grid"
    >
      <div
        className="row"
      >
        <div
          className="twelve wide column"
        >
          <input />
        </div>
        <div
          className="four wide column"
        >
          <a
            href={undefined}
            onClick={[Function]}
          >
            <button
              className="ui fluid button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
              type="button"
            >
              View
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field"
  >
    <label>
      Description
    </label>
    <textarea
      placeholder="Enter your description here (Markdown supported)"
    />
  </div>
  <div
    className="field"
  >
    <label>
      Preview
    </label>
    <div>
      Preview
    </div>
  </div>
  <div
    className="field"
  >
    <label />
    <div />
  </div>
  <div
    className="field"
  >
    <label>
      References 
      <i
        aria-hidden="true"
        className="info circle icon"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      />
    </label>
    <div>
      References
    </div>
  </div>
  <div
    className="inline field"
  >
    <label>
      Collaborators
    </label>
  </div>
</form>
`;

exports[`edit page form layouts should not contain slots if they were not provided 1`] = `
<form
  action={undefined}
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label>
      Title
    </label>
    <div
      className="ui divided two column grid"
    >
      <div
        className="row"
      >
        <div
          className="twelve wide column"
        >
          <input />
        </div>
        <div
          className="four wide column"
        >
          <a
            href={undefined}
            onClick={[Function]}
          >
            <button
              className="ui fluid button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
              type="button"
            >
              View
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field"
  >
    <label>
      Description
    </label>
    <textarea
      placeholder="Enter your description here (Markdown supported)"
    />
  </div>
  <div
    className="field"
  >
    <label />
    <div />
  </div>
  <div
    className="inline field"
  >
    <label>
      Collaborators
    </label>
  </div>
</form>
`;
